<?xml version="1.0" encoding="UTF-8"?>

<launch>
  <!-- Input -->
  <arg name="imu_in_topic" default="/stork/imu/data_raw" />
  <arg name="odom_in_topic" default="/stork/realsense/odom/sample" />
  <arg name="odom_in_topic2" default="/stork/livox/odometry" />
  <arg name="point_in_topic" default="/stork/leica/position" />

  <!-- Output Core State -->
  <arg name="full_state_out_topic" default="~full_state_out" />
  <arg name="full_state_lite_out_topic" default="~full_state_lite_out" />
  <arg name="pose_state_out_topic" default="~pose_state_out" />
  <arg name="odom_state_out_topic" default="~odom_state_out" />
  <arg name="path_states_out_topic" default="~path_states_out" />

  <!-- Output Calibration State -->
  <arg name="pose_cal_state_out_topic" default="~pose_cal_state_out" />
  <arg name="pose2_cal_state_out_topic" default="~pose2_cal_state_out" />
  <arg name="position_cal_state_out_topic" default="position_cal_state_out" />

  <!-- Node Options -->
  <arg name="respawn" default="false" />
  <arg name="rviz"    default="true" />

  <!-- Config File -->
  <arg name="config_file" default="$(find mars_ros)/launch/config/dualpose_position_config.yaml" />

  <!-- Launching the mars_ros pose_node -->
  <node name="mars_dualpose_position_node" pkg="mars_ros" type="dualpose_position_node" clear_params="true" output="screen" respawn="$(arg respawn)">

    <!-- Config File -->
    <rosparam command="load" file="$(arg config_file)" />

    <!-- Input -->
    <remap from="~imu_in" to="$(arg imu_in_topic)"/>
    <remap from="~odom_in" to="$(arg odom_in_topic)"/>
    <remap from="~odom2_in" to="$(arg odom_in_topic2)"/>
    <remap from="~point_in" to="$(arg point_in_topic)"/>

    <!-- Output Core State -->
    <remap from="~core_ext_state_out" to="$(arg full_state_out_topic)"/>
    <remap from="~core_ext_state_lite_out" to="$(arg full_state_lite_out_topic)"/>
    <remap from="~core_pose_state_out" to="$(arg pose_state_out_topic)"/>
    <remap from="~core_odom_state_out" to="$(arg odom_state_out_topic)"/>
    <remap from="~core_states_path" to="$(arg path_states_out_topic)"/>

    <!-- Output Calibration State -->
    <remap from="~pose_cal_state_out" to="$(arg pose_cal_state_out_topic)"/>
    <remap from="~pose2_cal_state_out" to="$(arg pose2_cal_state_out_topic)"/>
    <remap from="~position_cal_state_out" to="$(arg position_cal_state_out_topic)" />

  </node>

  <!-- launch rviz -->
  <group if="$(arg rviz)">
    <include file="$(find mars_ros)/launch/rviz/mars_rviz.launch">
      <arg name="topic_path_in"   value="$(arg path_states_out_topic)" />
      <arg name="topic_pose_in"   value="$(arg pose_state_out_topic)" />
      <arg name="namespace_mars"  value="mars_dualpose_position_node" />
    </include>

    <arg name="pose_topic"  value="$(eval eval(compile('pose_state_out_topic.replace(\'~\',\'\')','string','eval')))" />
    <node name="rviz_mars_pose_br" pkg="mars_ros" type="mars_tf2_converter" clear_params="true" output="screen" args="-t pose">
      <param name="child_id"      value="imu" />

      <remap from="~pose_in" to="$(arg pose_state_out_topic)" unless="$(eval pose_state_out_topic[0]=='~')"/>
      <remap from="~pose_in" to="/mars_dualpose_position_node/$(arg pose_topic)" if="$(eval pose_state_out_topic[0]=='~')"/>
    </node>

    <arg name="pose_cal_topic"  value="$(eval eval(compile('pose_cal_state_out_topic.replace(\'~\',\'\')','string','eval')))" />
    <node name="rviz_mars_pose_cal_br" pkg="mars_ros" type="mars_tf2_converter" clear_params="true" output="screen" args="-t pose">
      <param name="child_id"      value="rs" />
      <param name="parent_id"     value="imu" />

      <remap from="~pose_in" to="$(arg pose_cal_state_out_topic)" unless="$(eval pose_cal_state_out_topic[0]=='~')"/>
      <remap from="~pose_in" to="/mars_dualpose_position_node/$(arg pose_cal_topic)" if="$(eval pose_cal_state_out_topic[0]=='~')"/>
    </node>

    <arg name="pose2_cal_topic"  value="$(eval eval(compile('pose2_cal_state_out_topic.replace(\'~\',\'\')','string','eval')))" />
    <node name="rviz_mars_pose2_cal_br" pkg="mars_ros" type="mars_tf2_converter" clear_params="true" output="screen" args="-t pose">
      <param name="child_id"      value="livox" />
      <param name="parent_id"     value="imu" />

      <remap from="~pose_in" to="$(arg pose2_cal_state_out_topic)" unless="$(eval pose2_cal_state_out_topic[0]=='~')"/>
      <remap from="~pose_in" to="/mars_dualpose_position_node/$(arg pose2_cal_topic)" if="$(eval pose2_cal_state_out_topic[0]=='~')"/>
    </node>

    <arg name="pos_cal_topic"  value="$(eval eval(compile('position_cal_state_out_topic.replace(\'~\',\'\')','string','eval')))" />
    <node name="rviz_mars_pos_cal_br" pkg="mars_ros" type="mars_tf2_converter" clear_params="true" output="screen" args="-t pose">
      <param name="child_id"      value="prism" />
      <param name="parent_id"     value="imu" />

      <remap from="~pose_in" to="$(arg position_cal_state_out_topic)" unless="$(eval position_cal_state_out_topic[0]=='~')"/>
      <remap from="~pose_in" to="/mars_dualpose_position_node/$(arg pos_cal_topic)" if="$(eval position_cal_state_out_topic[0]=='~')"/>
    </node>
  </group>

</launch>
